cpp/src/Ice/ConnectionI.cpp

```diff
@@ -1400,10 +1397,12 @@ Ice::ConnectionI::message(ThreadPoolCurrent& current)
             {
                 assert(_state <= StateClosingPending);
 
-                //
-                // We parse messages first, if we receive a close
-                // connection message we won't send more messages.
-                //
+                if (wasAtRest && !isAtRest())
```

Comment:
In particular, how is it correct with concurrent writes?
For example, wasAtRest could be false because we were sending a oneway request; and once the sending is done, we scheduled the inactivity timer task. Now we receive a new frame (I think that's happening here) but don't cancel the inactivity timer task because wasAtRest is false?

Reply:
> In particular, how is it correct with concurrent writes?

What do you mean by "concurrent writes"? This code is protected by the connection lock so no writing/reading can be performed concurrently.

Reply:
Reading or writing on the transceiver is always performed with the mutex locked. You can't have concurrent writes while reading. In other words, a transceiver implementation can assume that read and write are never called concurrently.

The logic of `message` is really as follow:
1. reads and/or writes a message
2. if after 2, the rest state is  "it's no longer at rest" (`wasAtRest && !atRest()`) , cancel the inactivity timer
3. perform `parseMessage` and/or `sendNextMessage`
4. if after 3, the rest state is "it's now at rest" (`!wasAtReset && atRest()`), schedule the inactivity timer

"1" might or might not change the at rest state. It doesn't change it if reading/writing a connection validation message. It does change it otherwise.